marshal/consumer.go

```diff
@@ -518,11 +518,40 @@ func (c *Consumer) Terminate(release bool) bool {
 		}
 	}
 
+	// Optionally remove consumer from its marshal. Doing so is recommended
+	// if the marshal doesn't explicitly remove the consumer.
+	if remove {
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			c.marshal.lock.Lock()
```

Comment:
A few comments:

1. Reaching in to `c.marshal.lock` from the Consumer is, while technically legal, I feel a bit concerning as it breaks the sanctity of the interface. How about create a `func (m *Marshaler) removeConsumer(c *Consumer) error` on the Marshaler that the Consumer can call?

2. All lock acquisitions, with rare exception, must be of the form:

```
foo.Lock()
defer foo.Unlock()
```

I.e., we don't really allow splitting the lock/unlock away from each other as it leads to potential deadlocks and code that is harder to reason about.

3. I'm a bit confused about using the WaitGroup here. Couldn't you instead just use the closure synchronously, i.e., without the `go`? It just looks like you're using the WaitGroup to figure out when the goroutine exits, but since you're calling `wg.Wait` immediately after the closure, it looks like it's just a hack to get around the `go` running the closure asynchronously?